File: RyuHeySu/그래프/1446.py
# 지름길
import heapq
INF = int(1e9)

n, d = map(int, input().split())
graphs = [[] for i in range(d + 1)]
distance = [INF] * (d + 1)

for i in range(d) :
    graphs[i].append((i + 1, 1))
    
for _ in range(n) :
    start, end, length = map(int,input().split())
    
    if end > d :
        continue
    graphs[start].append((end, length))

def dijkastra(start) :
    q = []

    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q :
        dist, now = heapq.heappop(q) # 제일 짧은 교차로.

        # 이미 처리 되어 있음
        if distance[now] < dist :
            continue

        for node in graphs[now] :
            cost = dist + node[1]

            if distance[node[0]] > cost : # 이전 경로보다 현재 경로가 더 짧다면.
                distance[node[0]] = cost # 갱신
                heapq.heappush(q, (cost, node[0]))

dijkastra(0)
print(distance[d])

File: HeoJiye/beakjoon/graph/1446.js
// https://www.acmicpc.net/problem/1446

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "../test.txt";
const input = fs.readFileSync(filePath).toString();
const lines = input.split("\n");

const [N, D] = lines.shift().split(" ").map(Number);
const graph = {};
for (let i = 0; i <= 10000; i++) graph[i] = [];
lines.forEach((e) => {
  const [start, end, dist] = e.split(" ").map(Number);
  graph[start].push([end, dist]);
  if (end < D) graph[end].push([D, end - D]);
});
graph[0].push([D, D]);

const distance = [];
const INF = Number.MAX_SAFE_INTEGER;
for (let i = 0; i <= 10000; i++) distance[i] = INF;

function dijkstra(start, graph) {
  const heap = [];
  heap.push([0, start]);
  distance[start] = 0;

  while (heap.length > 0) {
    const [distFromStart, now] = heap.shift();
    if (distance[now] < distFromStart) continue;

    for (const [end, distFromNow] of graph[now]) {
      const cost = distFromStart + distFromNow;
      if (cost < distance[end]) {
        distance[end] = cost;
        heap.push([cost, end]);
        heap.sort((a, b) => a[0] - b[0]);
      }
    }
  }
}

dijkstra(0, graph);
console.log(distance[D]);
